### ISSUE TITLE
Terraform output will always return successful 

### ISSUE THREAD

**Issue by tobias-farrenkopf** [hashicorp/terraform#25956]
### Terraform Version

```
Terraform v0.12.28
Terraform v0.13.0
```

Terraform output will always return successfully which is rather unexpected.

### Expected Behavior

In case of a invalid or non existent output definition I expect an error message (check) and a return value != 0
### Actual Behavior

It will print a error statement on stderr but return successful (return code 0)
### Steps to Reproduce

Just run `terraform output nonsense` in a random folder

**Comment by alisdair** [comment]
This was an intentional behaviour change made in #23008, which I don't think we'll want to reverse.

Can you explain what the negative consequences of this behaviour are for you? Perhaps there's a workaround, or we can find an improvement to make here.

**Comment by tobias-farrenkopf** [comment]
Well it is at least a very unusual and unexpected behavior of a unix tool. Return success and print an error. We ran into an error, because we trusted the return value. Now I have to do ugly tricks like checking if something was printed on STDERR or check if a state file exists.
Regarding #23008 I have the feeling you optimized here for a corner case.

**Comment by alisdair** [comment]
I understand that it's unexpected, and I don't want you to think we're not interested in addressing that. But in order to improve the tool, we need to know more about the situation in which you're using `terraform output` and how this behaviour makes that difficult.

Can you tell us more about your desired workflow?

**Comment by tobias-farrenkopf** [comment]
Hi @alisdair,
we're using terraform in a bigger deployment scope. So basically other components need to now urls and other stuff which we deployed with the help of terraform. Therefor we use `terraform output` to extract things like ip addreses etc... from other bash/python files. Of course everything is working fine as long as nothing goes wrong and your output variables are set.
Our assumption was, that a missing state file would trigger an error. Instead we had to debug a lot to figure that one out.

**Comment by alisdair** [comment]
That makes sense, and thanks for explaining! 😄 

This is a tricky issue to figure out, because it's going to be a breaking change no matter what we do. One of the challenges here is that an empty or missing state file is not normally a source of errors in Terraform—it's a valid initial state, and not one that we consider to be a problem in most cases.

So at present, running `terraform output` with an empty state file only reports a UI-level error (rather than a shell error exit code) because it's also valid for a configuration to have no outputs. It's like running `ls` in an empty directory.

I have a feeling that there might be a reasonable proposal for how we could modify the `output` subcommand in this situation to be more universally helpful. Unfortunately I can't think of one, and I don't think making `output` return an error exit code when there's no state file is likely to be a good solution. I'm going to leave this issue open in the hope that someone can come up with a good proposal here.

Two notes that may be helpful to future people finding this issue and looking for workarounds:

- For automation and scripting, it might be better to use the `-json` flag to Terraform subcommands where possible. At the moment, the `output` subcommand does error on missing state file if you are looking for a specific output **and** you're using JSON output. Example:

    ```shellsession
    $ terraform output nonsense 2>/dev/null || echo error
    $ terraform output -json nonsense 2>/dev/null || echo error
    error
    ```

    The `-json` flag is intended for automation workflows such as the one you're building.

- Alternatively, the `state list` subcommand will error on missing state file, which you could use as a check to see if the state file is present:

    ```shellsession
    $ terraform state list >/dev/null 2>&1 || echo error
    error
    ```

**Comment by tobias-farrenkopf** [comment]
Well it is as you say: You throw an UI-error but it actually is none. Maybe you change the output to a warning or indicate in another way that you will return a success anyway.
The -json flag is a good approach (at the moment we check if stderr is empty) and mitigates our problem.

Anyway. We're aware about this now. I would hope that you either improve the documentation or change the error message. But also feel free to close it.

**Comment by alisdair** [comment]
The fix for this has been merged and will be in the 0.13.2 release. Thanks for your patience in working with me to find a good solution to this surprisingly tricky issue!
